### PR TITLE
Add tx pool to firehose block

### PR DESCRIPTION
**Pull request by mariusmihaic**
## Reasoning behind the pull request
- Include tx pool in firehose block

## Proposed changes
- Add support to parse tx pool and convert each interface to its proto message

## Testing procedure
- Local testnet + firehose ingestion process + tools to check firehose ingested blocks

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: outport/firehose/firehoseIndexerTxPool.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	smartContractResult map[string]*firehose.SCRWithFee
+	rewards             map[string]*rewardTx.RewardTx
+	receipts            map[string]*receipt.Receipt
+	invalidTxs          map[string]*firehose.TxWithFee
+	logs                map[string]*transaction.Log
+}
+
+func getTxPool(transactionsPool *outportcore.Pool) (*txPool, error) {
+	if transactionsPool == nil {
+		return &txPool{}, nil
```
Comment (SebastianMarian) [inline outport/firehose/firehoseIndexerTxPool.go:26]: So if the given transactionPool is nil we will return an empty txPool DTO?

File: outport/firehose/firehoseIndexerTxPool.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+		transactions:        txs,
+		smartContractResult: scrs,
+		rewards:             rewards,
+		receipts:            receipts,
+		invalidTxs:          invalidTxs,
+		logs:                logs,
+	}, nil
+}
+
+func getFirehoseFreeInfo(txHandler data.TransactionHandlerWithGasUsedAndFee) *firehose.FeeInfo {
```
Comment (AdoAdoAdo) [inline outport/firehose/firehoseIndexerTxPool.go:64]: *free -> fee

File: outport/firehose/firehoseIndexer.go [hunk 3]
```diff
@@ -105,6 +100,28 @@ func (fi *firehoseIndexer) SaveBlock(args *outportcore.ArgsSaveBlockData) error
 	return nil
 }
 
+func (fi *firehoseIndexer) getHeaderBytes(headerHandler data.HeaderHandler) ([]byte, core.HeaderType, error) {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply outport/firehose/firehoseIndexer.go:103]: This is already tested by `TestFirehoseIndexer_SaveBlockHeader` from `firehoseIndexer_test.go`

File: outport/firehose/firehoseIndexerTxPool.go (hunk 2 above)
Comment (mariusmihaic) [reply outport/firehose/firehoseIndexerTxPool.go:64]: Right, good catch

File: outport/firehose/firehoseIndexerTxPool.go (hunk 1 above)
Comment (mariusmihaic) [reply outport/firehose/firehoseIndexerTxPool.go:26]: Yes, that's desired. Wanted a logic to either fill the fields or data, or, if none available -> nil

**Review by AdoAdoAdo — DISMISSED** [review]

File: outport/firehose/firehoseIndexer_test.go [hunk 4]
```diff
@@ -14,7 +14,7 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestFirehoseIndexer_SaveBlock(t *testing.T) {
+func TestFirehoseIndexer_SaveBlockHeader(t *testing.T) {
```
Comment (raduchis) [inline outport/firehose/firehoseIndexer_test.go:17]: please add an unit test for the constructor with nil writer.

File: outport/firehose/firehoseIndexerTxPool.go (hunk 1 above)
Comment (raduchis) [reply outport/firehose/firehoseIndexerTxPool.go:26]: the indexing could be inconsistent: the header should have transactions but because the transactionsPool is nil, the txs, scrs etc, will not be indexed. Is there no way to verify the consistency?

File: outport/firehose/firehoseIndexerTxPool_test.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+	ioWriterCalledCt := 0
+	ioWriter := &testscommon.IoWriterStub{
+		WriteCalled: func(p []byte) (n int, err error) {
+			ioWriterCalledCt++
+			switch ioWriterCalledCt {
+			case 1:
+				require.Equal(t, []byte("FIRE BLOCK_BEGIN 2\n"), p)
+			case 2:
+
```
Comment (raduchis) [inline outport/firehose/firehoseIndexerTxPool_test.go:221]: empty line could be removed

**Review by raduchis — CHANGES_REQUESTED** [review]

File: outport/firehose/firehoseIndexerTxPool.go (hunk 1 above)
Comment (mariusmihaic) [reply outport/firehose/firehoseIndexerTxPool.go:26]: I don't think there could be inconsistencies here, since txs are saved in outport driver from current saved block:

```
return &outportcore.Pool{
   Txs:      WrapTxsMap(odp.txCoordinator.GetAllCurrentUsedTxs(block.TxBlock)),
   Scrs:     WrapTxsMap(odp.txCoordinator.GetAllCurrentUsedTxs(block.SmartContractResultBlock)),
   Rewards:  WrapTxsMap(odp.txCoordinator.GetAllCurrentUsedTxs(block.RewardsBlock)),
   Invalid:  WrapTxsMap(odp.txCoordinator.GetAllCurrentUsedTxs(block.InvalidBlock)),
   Receipts: WrapTxsMap(odp.txCoordinator.GetAllCurrentUsedTxs(block.ReceiptBlock)),
   Logs:     odp.txCoordinator.GetAllCurrentLogs(),
}
```

Comment (mariusmihaic) [reply outport/firehose/firehoseIndexerTxPool.go:26]: However, as discussed, changed logic to return error in case of nil tx pool

File: outport/firehose/firehoseIndexer_test.go (hunk 4 above)
Comment (mariusmihaic) [reply outport/firehose/firehoseIndexer_test.go:17]: Sure, added `TestNewFirehoseIndexer`

File: outport/firehose/firehoseIndexerTxPool_test.go (hunk 5 above)
Comment (mariusmihaic) [reply outport/firehose/firehoseIndexerTxPool_test.go:221]: Right, removed it from here and other places as well

File: outport/firehose/firehoseIndexer_test.go (hunk 4 above)
Comment (raduchis) [reply outport/firehose/firehoseIndexer_test.go:37]: ok

**Review by raduchis — APPROVED** [review]

File: outport/firehose/firehoseIndexerTxPool_test.go (hunk 5 above)
Comment (raduchis) [reply outport/firehose/firehoseIndexerTxPool_test.go:238]: ok

File: outport/firehose/firehoseIndexer_test.go (hunk 4 above)
Comment (raduchis) [reply outport/firehose/firehoseIndexer_test.go:37]: ok
